lib/Conversion/CalyxToFSM/CalyxToFSM.cpp

```diff
@@ -268,6 +268,71 @@ LogicalResult CompileFSMVisitor::visit(StateOp currentState, EnableOp enableOp,
   return success();
 }
 
+// CompileInvoke is used to convert invoke operations to group operations and
+// enable operations.
+class CompileInvoke {
+public:
+  CompileInvoke(ComponentOp component, OpBuilder builder)
+      : component(component), builder(builder) {}
+  void compile();
+
+private:
+  void lowerInvokeOp(InvokeOp invokeOp);
+  ComponentOp component;
+  OpBuilder builder;
+  // It is used to pass to the go port and assign a value to the go port.
+  hw::ConstantOp constantOp = nullptr;
+  llvm::StringMap<size_t> groupNameMap;
+};
+
+// Access all invokeOp.
+void CompileInvoke::compile() {
+  llvm::SmallVector<InvokeOp> invokeOps =
+      component.getControlOp().getInvokeOps();
+  for (InvokeOp op : invokeOps) {
+    lowerInvokeOp(op);
+  }
+}
+
+// Convert an invoke operation to a group operation and an enable operation.
+void CompileInvoke::lowerInvokeOp(InvokeOp invokeOp) {
+  // Create a ConstantOp to assign a value to the go port.
+  if (!constantOp) {
+    Operation *prevNode = component.getWiresOp().getOperation()->getPrevNode();
+    builder.setInsertionPointAfter(prevNode);
+    constantOp = builder.create<hw::ConstantOp>(prevNode->getLoc(),
+                                                builder.getI1Type(), 1);
+  }
+  Location loc = component.getWiresOp().getLoc();
+  // Set the insertion point at the end of the wires block.
+  builder.setInsertionPointToEnd(component.getWiresOp().getBodyBlock());
+  llvm::StringRef callee = invokeOp.getCallee();
+  if (groupNameMap.find(callee) == groupNameMap.end())
+    groupNameMap[callee] = 0;
```

Comment:
You're doing many look ups here. Save `groupNameMap[callee]` (or its iterator) to a local variable.